# Aufgabe 15 - SENet

Dieses Notebook thematisiert den Forward und Backward Pass durch einen (SE-)ResNet-v1-Bottleneck-Block.

Ziel ist es, den Forward und Backward Pass in PyTorch zu realisieren und die Ergebnisse aus den Übungsunterlagen zu reproduzieren.

### Inhaltsverzeichnis
- [(d) Implementierung des SE-ResNet-v1-Bottleneck-Blocks in PyTorch](#d)
    - [Implementierung des Squeeze-and-Excitation-Blocks](#se-block)
    - [Implementierung des Bottleneck-Blocks](#bottleneck-block)
    - [Block erstellen und Forward und Backward Propagation anwenden](#se-anwendung)
- [Reproduktion der Ergebnisse aus Aufgabe 14](#reproduktion-a14)

<hr style="border-width: 5px">

### Vorbereitung
Der Übersicht halber sind einige Funktionalitäten in ein separates Paket ausgelagert. Grundvoraussetzung für deren Verwendung ist, dass Sie das Paket `tui-dl4cv` <font color="#aa0000">installieren bzw. aktualisieren</font> und anschließend importieren.

Für die Installation stehen Ihnen zwei mögliche Wege zur Verfügung.

**(1) Installation direkt in diesem Notebook:**
Führen Sie den nachfolgenden Code-Block aus.

In [1]:
import sys

print(f"Automatically install package for '{sys.executable}'")
!{sys.executable} -m pip install tui-dl4cv \
    --extra-index-url "https://2022ws:xXCgQHZxxeNYchgryN7e@nikrgl.informatik.tu-ilmenau.de/api/v4/projects/1730/packages/pypi/simple" \
    --no-cache --upgrade

Automatically install package for '/usr/bin/python3'
Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://pypi.org/simple, https://2022ws:****@nikrgl.informatik.tu-ilmenau.de/api/v4/projects/1730/packages/pypi/simple
DEPRECATION: The HTML index page being used (https://nikrgl.informatik.tu-ilmenau.de/api/v4/projects/1730/packages/pypi/simple/tui-dl4cv/) is not a proper HTML 5 document. This is in violation of PEP 503 which requires these pages to be well-formed HTML 5 documents. Please reach out to the owners of this index page, and ask them to update this index page to a valid HTML 5 document. pip 22.2 will enforce this behaviour change. Discussion can be found at https://github.com/pypa/pip/issues/10825


ODER

**(2) Manuelle Installation über die Konsole:**
Öffnen Sie eine Konsole ("Anaconda Prompt" unter Windows) und führen Sie folgenden Befehl aus:
```text
pip install tui-dl4cv --extra-index-url "https://2022ws:xXCgQHZxxeNYchgryN7e@nikrgl.informatik.tu-ilmenau.de/api/v4/projects/1730/packages/pypi/simple" --no-cache --upgrade
```

**Führen Sie abschließend folgenden Code-Block aus, um das Paket verwenden zu können.**
Während der Bearbeitung können Sie nun Ihre Ergebnisse mithilfe der Funktion `interactive_check` überprüfen. Die Funktionsaufrufe sind bereits an den entsprechenden Stellen im Notebook enthalten.

In [2]:
import tui_dl4cv.seresnet

# noetige Erweiterung, damit Variablen aus diesem Notebook automatisch ueberprueft werden koennen
def interactive_check(name, **kwargs):
    tui_dl4cv.seresnet.interactive_check(name, globals(), **kwargs)

from tui_dl4cv.seresnet import apply_block_weights
from tui_dl4cv.seresnet import print_tensors

<hr style="border-width: 5px">

<a name="d"></a>
### (d) Vollziehen Sie die Backpropagation durch den SE-ResNet-v1-Bottleneck-Block anhand einer PyTorch-Implementierung nach.

---
Pakete importieren:

In [3]:
# PyTorch
import torch

---

<a name="se-block"></a>
*Squeeze-and-Excitation-Block implementieren:*

Zunächst soll eine PyTorch-Klasse für den Squeeze-and-Excitation-Block implementiert werden.

Im Squeeze-and-Excitation-Block wird zunächst mittels Global Average Pooling die Räumlichkeit entfernt (Squeeze-Operation). Anschließend werden mithilfe von zwei Convolutions (Excitation-Operation) Wichtungsfaktoren für jeden Kanal berechnet. In der Excitation-Operation wird die Kanalanzahl dabei zuerst um den Faktor `r` reduziert, um eine *fokusierte Anregunng* zu bestimmen, und anschließend wieder auf die ursprüngliche Kanalanzahl zurückprojiziert.

<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
Folgende PyTorch-Definitionen könnten für die Vervollständigung der Lücken hilfreich sein:
    <ul style="margin-bottom: 0px">
        <li><code style="background-color: #FAEAEA;">torch.nn.Module</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.nn.Module.html" target="_blank">PyTorch-Dokumentation</a>
        </li>
        <li><code style="background-color: #FAEAEA;">torch.nn.Conv2d</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html" target="_blank">PyTorch-Dokumentation</a>
        </li>
    </ul>
</div>

In [4]:
class SqueezeAndExcitationBlock(torch.nn.Module):
    def __init__(self, n_channels, r=16):
        super().__init__()

        # Schichten anlegen
        self.conv1 = torch.nn.Conv2d(
        in_channels=n_channels,
        out_channels=int(n_channels/r),
        kernel_size=1, bias=True
        )    # bitte Code ergaenzen <---------------- [Luecke (1)]
        self.conv2 = torch.nn.Conv2d(
        in_channels=int(n_channels/r),
        out_channels=n_channels,
        kernel_size=1,bias=True
        )    # bitte Code ergaenzen <---------------- [Luecke (2)]

    def forward(self, x):
        # Kopieren des Inputs, damit die Gradienten der einzelnen Pfade
        # getrennt voneinander ausgegeben werden koennen
        x_skip = x.clone()
        x_se = x.clone()

        # Squeeze (Global Average Pooling = Mittelwert ueber Dim. 2 und 3)
        gap_out =  x_se.mean(dim=[2,3],keepdim=True)   # bitte Code ergaenzen <---------------- [Luecke (3)]

        # Excitation
        conv1_act = self.conv1(gap_out)    # bitte Code ergaenzen <---------------- [Luecke (4)]
        conv1_out = torch.relu(conv1_act)    # bitte Code ergaenzen <---------------- [Luecke (5)]
        conv2_act = self.conv2(conv1_out) #bitte Code ergaenzen <---------------- [Luecke (6)]
        conv2_out = torch.sigmoid(conv2_act)    # bitte Code ergaenzen <---------------- [Luecke (7)]

        # Skalierung der Eingangswerte
        y = conv2_out * x_skip    # bitte Code ergaenzen <---------------- [Luecke (8)]

        # Ergebnisse fuer Forward Pass ausgeben
        print_tensors((gap_out, conv1_act, conv1_out, conv2_act, conv2_out, y),
                      ('se_gap_out', 'se_conv1_act', 'se_conv1_out', 'se_conv2_act', 'se_conv2_out', 'se_y'))

        # Ergebnisse fuer Backward Pass ausgeben
        # Realisierung ueber Hooks
        # hier: nur Gradienten des Vordrucks
        gap_out.register_hook(lambda grad: print_tensors(grad, 'se_gap_out.grad'))
        conv2_out.register_hook(lambda grad: print_tensors(grad, 'se_conv2_out.grad'))
        x_skip.register_hook(lambda grad: print_tensors(grad, 'se_x_skip.grad'))
        x_se.register_hook(lambda grad: print_tensors(grad, 'se_x_se.grad'))
        x.register_hook(lambda grad: print_tensors(grad, 'se_x.grad'))

        return y

---

<a name="bottleneck-block"></a>
*(SE-)ResNet-v1-Bottleneck-Block implementieren:*

Im nächsten Schritt soll der (SE-)ResNet-v1-Bottleneck-Block implementiert werden.
Die Verwendung des Squeeze-and-Excitation-Blocks soll dabei optional sein, damit später die Ergebnisse für beide Fälle reproduziert werden können.

Im Bottleneck-Block wird die Anzahl der Input Feature Maps (`n_channels_in`) zunächst mithilfe einer ersten 1x1 Convolution auf `n_channels` reduziert. Anschließend folgt eine 3x3 Convolution, bei der die Kanalanzahl unverändert bleibt. Abschließend wird die Anzahl der Kanäle mithilfe einer weiteren 1x1 Convolution um den Faktor 4 expandiert.

<br>
<div style="background-color: #EAF2F8; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
Beachten Sie, dass in der nachfolgenden Implementierung bewusst auf Batch Normalization verzichtet wird, um Ergebnisgleichheit mit der Rechenaufgabe zu erreichen. Die entsprechenden Stellen sind daher auskommentiert. Weiterhin unterstützt die Implementierung kein Downsampling. Die Skip Connection ist daher immer frei von zusätzlichen Operationen. Beide Aspekte werden in der nächsten Übungsaufgabe aufgegriffen.
</div>

<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
Folgende PyTorch-Definitionen könnten für die Vervollständigung der Lücken hilfreich sein:
    <ul style="margin-bottom: 0px">
        <li><code style="background-color: #FAEAEA;">torch.nn.Module</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.nn.Module.html" target="_blank">PyTorch-Dokumentation</a>
        </li>
        <li><code style="background-color: #FAEAEA;">torch.nn.Identity</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.nn.Identity.html" target="_blank">PyTorch-Dokumentation</a>
        </li>
        <li><code style="background-color: #FAEAEA;">torch.nn.Conv2d</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html" target="_blank">PyTorch-Dokumentation</a>
        </li>
        <li><code style="background-color: #FAEAEA;">torch.nn.BatchNorm2d</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html" target="_blank">PyTorch-Dokumentation</a>
        </li>
    </ul>
</div>

In [5]:
class BottleneckBlock(torch.nn.Module):

    expansion = 4

    def __init__(self, n_channels_in, n_channels, r_se=None, stride=1):
        super().__init__()

        # keine Unterstuetzung fuer Downsampling bzw. Aenderung der Kanalanzahl
        assert stride == 1
        self.skip = torch.nn.Identity()

        # Schichten anlegen
        self.conv1 = torch.nn.Conv2d(
        in_channels=n_channels_in,
        out_channels=n_channels,
        kernel_size=1, stride=1, bias=False
        )   # bitte Code ergaenzen <---------------- [Luecke (9)]
        # self.conv1_bn = torch.nn.BatchNorm2d(n_channels)

        self.conv2 = torch.nn.Conv2d(
        in_channels=n_channels,
        out_channels=n_channels,
        kernel_size=3, stride=1, padding=1, bias=False
        )    # bitte Code ergaenzen <---------------- [Luecke (10)]
        # self.conv2_bn = torch.nn.BatchNorm2d(n_channels)

        self.conv3 = torch.nn.Conv2d(
        in_channels=n_channels,
        out_channels=n_channels*self.expansion,
        kernel_size=1, stride=1, bias=False
        )    # bitte Code ergaenzen <---------------- [Luecke (11)]
        # self.conv3_bn = torch.nn.BatchNorm2d(channels_out)

        # Squeeze-and-Excitation-Block
        if r_se is not None:
            # SE soll verwendet werden
            self.se = SqueezeAndExcitationBlock(
            n_channels=n_channels*self.expansion,
            r=r_se
            )    # bitte Code ergaenzen <---------------- [Luecke (12)]
        else:
            self.se = None

    def forward(self, x):
        # Kopieren des Inputs, damit die Gradienten der einzelnen Pfade
        # getrennt voneinander ausgegeben werden koennen
        x_skip = x.clone()
        x_res = x.clone()

        # Residuum
        conv1_act = self.conv1(x_res)    # bitte Code ergaenzen <---------------- [Luecke (13)]
        # conv1_act = self.conv1_bn(conv1_act)
        conv1_out = torch.relu(conv1_act)   # bitte Code ergaenzen <---------------- [Luecke (14)]

        conv2_act =  self.conv2(conv1_out)   # bitte Code ergaenzen <---------------- [Luecke (15)]
        # conv2_act = self.conv2_bn(conv2_act)
        conv2_out =  torch.relu(conv2_act)    # bitte Code ergaenzen <---------------- [Luecke (16)]

        conv3_act =  self.conv3(conv2_out)    # bitte Code ergaenzen <---------------- [Luecke (17)]
        # conv3_act = self.conv3_bn(conv3_act)

        # Ergebnisse fuer Forward Pass ausgeben
        print_tensors((conv1_act, conv1_out, conv2_act, conv2_out, conv3_act),
                      ('conv1_act', 'conv1_out', 'conv2_act', 'conv2_out', 'conv3_act'))

        # SE
        if self.se is not None:
            res = self.se(conv3_act)    # bitte Code ergaenzen <---------------- [Luecke (18)]
        else:
            res = conv3_act    # bitte Code ergaenzen <---------------- [Luecke (19)]

        # Skip Connection
        skip = self.skip(x_skip)

        # beide Zweige zusammenfuehren
        z =  res+skip   # bitte Code ergaenzen <---------------- [Luecke (20)]
        y =  torch.relu(z)   # bitte Code ergaenzen <---------------- [Luecke (21)]

        # Ergebnisse fuer Forward Pass ausgeben
        print_tensors((res, z, y), ('res', 'z', 'y'))

        # Ergebnisse fuer Backward Pass ausgeben
        # Realisierung ueber Hooks
        # hier: nur Gradienten des Vordrucks
        z.register_hook(lambda grad: print_tensors(grad, 'z.grad'))
        res.register_hook(lambda grad: print_tensors(grad, 'res.grad'))
        x_skip.register_hook(lambda grad: print_tensors(grad, 'x_skip.grad'))
        x_res.register_hook(lambda grad: print_tensors(grad, 'x_res.grad'))
        x.register_hook(lambda grad: print_tensors(grad, 'x.grad'))

        return y

---

<a name="se-anwendung"></a>
*SE-ResNet-v1-Bottleneck-Block erstellen und Gewichte laden:*

Der Übersicht halber sind die Gewichte bereits im `tui-dl4cv`-Paket definiert und können über die Funktion `apply_block_weights` geladen werden.

In [6]:
# Block-Objekt erzeugen
block = BottleneckBlock(
    n_channels_in=8,
    n_channels=2,
    r_se=4
)

# Gewichte laden (nur fuer oben definierten Block)
apply_block_weights(block)

---

*Forward Pass:*

In [7]:
# Eingabe definieren
x = torch.tensor(
    [[[[1, 1, 0],
       [0, 0, 0],
       [0, 1, 0]],
      [[1, 0, 0],
       [1, 1, 0],
       [0, 1, 0]],
      [[1, 0, 0],
       [0, 1, 0],
       [0, 1, 0]],
      [[1, 1, 0],
       [0, 1, 2],
       [0, 0, 0]],
      [[3, 0, 2],
       [0, 0, 1],
       [0, 1, 0]],
      [[0, 0, 0],
       [0, 0, 0],
       [1, 0, 1]],
      [[1, 0, 0],
       [0, 3, 0],
       [0, 1, 0]],
      [[0, 0, 0],
       [0, 0, 2],
       [1, 0, 1]]]],
    dtype=torch.float32, requires_grad=True
)

# Forward Propagation
y = block(x)

# Ergebnis ueberpruefen
interactive_check('seresnet_forward')

conv1_act:
[[[[ 1.  1.  2.]
   [ 2. -9.  5.]
   [ 3. -1.  3.]]

  [[ 2. -3.  2.]
   [ 0.  3.  1.]
   [ 1.  1.  1.]]]]
conv1_out:
[[[[1. 1. 2.]
   [2. 0. 5.]
   [3. 0. 3.]]

  [[2. 0. 2.]
   [0. 3. 1.]
   [1. 1. 1.]]]]
conv2_act:
[[[[ -1.  -5.  -9.]
   [ -1.   0.  -6.]
   [  9. -18.  10.]]

  [[ -5.  -1.  -1.]
   [ -1.  -4.  10.]
   [ 12. -14.   8.]]]]
conv2_out:
[[[[ 0.  0.  0.]
   [ 0.  0.  0.]
   [ 9.  0. 10.]]

  [[ 0.  0.  0.]
   [ 0.  0. 10.]
   [12.  0.  8.]]]]
conv3_act:
[[[[  0.   0.   0.]
   [  0.   0.  20.]
   [-30.   0. -44.]]

  [[  0.   0.   0.]
   [  0.   0. -10.]
   [ 15.   0.  22.]]

  [[  0.   0.   0.]
   [  0.   0.  30.]
   [ 36.   0.  24.]]

  [[  0.   0.   0.]
   [  0.   0.  10.]
   [-15.   0. -22.]]

  [[  0.   0.   0.]
   [  0.   0. -20.]
   [ 30.   0.  44.]]

  [[  0.   0.   0.]
   [  0.   0.   0.]
   [  0.   0.   0.]]

  [[  0.   0.   0.]
   [  0.   0.  20.]
   [ 51.   0.  46.]]

  [[  0.   0.   0.]
   [  0.   0.  50.]
   [ 87.   0.  70.]]]]
se_gap_out:
[[[[-6.]

Überprüfung der Ausgabe 'y':


<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
    <br>
    <i>Ausgaben bei korrekter Implementierung:</i>
    <br>
    <code style="padding: 0">
conv1_act:
[[[[ 1.  1.  2.]
   [ 2. -9.  5.]
   [ 3. -1.  3.]]
  [[ 2. -3.  2.]
   [ 0.  3.  1.]
   [ 1.  1.  1.]]]]
conv1_out:
[[[[1. 1. 2.]
   [2. 0. 5.]
   [3. 0. 3.]]
  [[2. 0. 2.]
   [0. 3. 1.]
   [1. 1. 1.]]]]
conv2_act:
[[[[ -1.  -5.  -9.]
   [ -1.   0.  -6.]
   [  9. -18.  10.]]
  [[ -5.  -1.  -1.]
   [ -1.  -4.  10.]
   [ 12. -14.   8.]]]]
conv2_out:
[[[[ 0.  0.  0.]
   [ 0.  0.  0.]
   [ 9.  0. 10.]]
  [[ 0.  0.  0.]
   [ 0.  0. 10.]
   [12.  0.  8.]]]]
conv3_act:
[[[[  0.   0.   0.]
   [  0.   0.  20.]
   [-30.   0. -44.]]
  [[  0.   0.   0.]
   [  0.   0. -10.]
   [ 15.   0.  22.]]
  [[  0.   0.   0.]
   [  0.   0.  30.]
   [ 36.   0.  24.]]
  [[  0.   0.   0.]
   [  0.   0.  10.]
   [-15.   0. -22.]]
  [[  0.   0.   0.]
   [  0.   0. -20.]
   [ 30.   0.  44.]]
  [[  0.   0.   0.]
   [  0.   0.   0.]
   [  0.   0.   0.]]
  [[  0.   0.   0.]
   [  0.   0.  20.]
   [ 51.   0.  46.]]
  [[  0.   0.   0.]
   [  0.   0.  50.]
   [ 87.   0.  70.]]]]
se_gap_out:
[[[[-6.]]
  [[ 3.]]
  [[10.]]
  [[-3.]]
  [[ 6.]]
  [[ 0.]]
  [[13.]]
  [[23.]]]]
se_conv1_act:
[[[[1.]]
  [[6.]]]]
se_conv1_out:
[[[[1.]]
  [[6.]]]]
se_conv2_act:
[[[[  0.]]
  [[ 20.]]
  [[ 12.]]
  [[ 13.]]
  [[  0.]]
  [[-13.]]
  [[ 19.]]
  [[-18.]]]]
se_conv2_out:
[[[[0.5]]
  [[1. ]]
  [[1. ]]
  [[1. ]]
  [[0.5]]
  [[0. ]]
  [[1. ]]
  [[0. ]]]]
se_y:
[[[[  0.   0.   0.]
   [  0.   0.  10.]
   [-15.   0. -22.]]
  [[  0.   0.   0.]
   [  0.   0. -10.]
   [ 15.   0.  22.]]
  [[  0.   0.   0.]
   [  0.   0.  30.]
   [ 36.   0.  24.]]
  [[  0.   0.   0.]
   [  0.   0.  10.]
   [-15.   0. -22.]]
  [[  0.   0.   0.]
   [  0.   0. -10.]
   [ 15.   0.  22.]]
  [[  0.   0.   0.]
   [  0.   0.   0.]
   [  0.   0.   0.]]
  [[  0.   0.   0.]
   [  0.   0.  20.]
   [ 51.   0.  46.]]
  [[  0.   0.   0.]
   [  0.   0.   0.]
   [  0.   0.   0.]]]]
res:
[[[[  0.   0.   0.]
   [  0.   0.  10.]
   [-15.   0. -22.]]
  [[  0.   0.   0.]
   [  0.   0. -10.]
   [ 15.   0.  22.]]
  [[  0.   0.   0.]
   [  0.   0.  30.]
   [ 36.   0.  24.]]
  [[  0.   0.   0.]
   [  0.   0.  10.]
   [-15.   0. -22.]]
  [[  0.   0.   0.]
   [  0.   0. -10.]
   [ 15.   0.  22.]]
  [[  0.   0.   0.]
   [  0.   0.   0.]
   [  0.   0.   0.]]
  [[  0.   0.   0.]
   [  0.   0.  20.]
   [ 51.   0.  46.]]
  [[  0.   0.   0.]
   [  0.   0.   0.]
   [  0.   0.   0.]]]]
z:
[[[[  1.   1.   0.]
   [  0.   0.  10.]
   [-15.   1. -22.]]
  [[  1.   0.   0.]
   [  1.   1. -10.]
   [ 15.   1.  22.]]
  [[  1.   0.   0.]
   [  0.   1.  30.]
   [ 36.   1.  24.]]
  [[  1.   1.   0.]
   [  0.   1.  12.]
   [-15.   0. -22.]]
  [[  3.   0.   2.]
   [  0.   0.  -9.]
   [ 15.   1.  22.]]
  [[  0.   0.   0.]
   [  0.   0.   0.]
   [  1.   0.   1.]]
  [[  1.   0.   0.]
   [  0.   3.  20.]
   [ 51.   1.  46.]]
  [[  0.   0.   0.]
   [  0.   0.   2.]
   [  1.   0.   1.]]]]
y:
[[[[ 1.  1.  0.]
   [ 0.  0. 10.]
   [ 0.  1.  0.]]
  [[ 1.  0.  0.]
   [ 1.  1.  0.]
   [15.  1. 22.]]
  [[ 1.  0.  0.]
   [ 0.  1. 30.]
   [36.  1. 24.]]
  [[ 1.  1.  0.]
   [ 0.  1. 12.]
   [ 0.  0.  0.]]
  [[ 3.  0.  2.]
   [ 0.  0.  0.]
   [15.  1. 22.]]
  [[ 0.  0.  0.]
   [ 0.  0.  0.]
   [ 1.  0.  1.]]
  [[ 1.  0.  0.]
   [ 0.  3. 20.]
   [51.  1. 46.]]
  [[ 0.  0.  0.]
   [ 0.  0.  2.]
   [ 1.  0.  1.]]]]
</code>
</details>

---

*Backward Pass:*

In [8]:
# Gradienten definieren
y_grad = torch.ones(y.shape)

# Backpass durchfuehren
y.backward(y_grad)

# Ergebnis ueberpruefen
interactive_check('seresnet_backward')

z.grad:
[[[[1. 1. 0.]
   [0. 0. 1.]
   [0. 1. 0.]]

  [[1. 0. 0.]
   [1. 1. 0.]
   [1. 1. 1.]]

  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]

  [[1. 1. 0.]
   [0. 1. 1.]
   [0. 0. 0.]]

  [[1. 0. 1.]
   [0. 0. 0.]
   [1. 1. 1.]]

  [[0. 0. 0.]
   [0. 0. 0.]
   [1. 0. 1.]]

  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]

  [[0. 0. 0.]
   [0. 0. 1.]
   [1. 0. 1.]]]]
res.grad:
[[[[1. 1. 0.]
   [0. 0. 1.]
   [0. 1. 0.]]

  [[1. 0. 0.]
   [1. 1. 0.]
   [1. 1. 1.]]

  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]

  [[1. 1. 0.]
   [0. 1. 1.]
   [0. 0. 0.]]

  [[1. 0. 1.]
   [0. 0. 0.]
   [1. 1. 1.]]

  [[0. 0. 0.]
   [0. 0. 0.]
   [1. 0. 1.]]

  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]

  [[0. 0. 0.]
   [0. 0. 1.]
   [1. 0. 1.]]]]
se_conv2_out.grad:
[[[[ 20.]]

  [[ 37.]]

  [[ 90.]]

  [[ 10.]]

  [[ 74.]]

  [[  0.]]

  [[117.]]

  [[207.]]]]
se_gap_out.grad:
[[[[ -44.99]]

  [[  89.98]]

  [[ -25.  ]]

  [[ -25.  ]]

  [[-119.99]]

  [[   0.  ]]

  [[  45.  ]]

  [[ -10.  ]]]]
se_x_se.grad:
[[[[ -5.

Überprüfung des Gradienten von 'x':


<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
    <br>
    <i>Augaben bei korrekter Implementierung:</i>
    <br>
    <code style="padding: 0">
z.grad:
[[[[1. 1. 0.]
   [0. 0. 1.]
   [0. 1. 0.]]
  [[1. 0. 0.]
   [1. 1. 0.]
   [1. 1. 1.]]
  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]
  [[1. 1. 0.]
   [0. 1. 1.]
   [0. 0. 0.]]
  [[1. 0. 1.]
   [0. 0. 0.]
   [1. 1. 1.]]
  [[0. 0. 0.]
   [0. 0. 0.]
   [1. 0. 1.]]
  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]
  [[0. 0. 0.]
   [0. 0. 1.]
   [1. 0. 1.]]]]
res.grad:
[[[[1. 1. 0.]
   [0. 0. 1.]
   [0. 1. 0.]]
  [[1. 0. 0.]
   [1. 1. 0.]
   [1. 1. 1.]]
  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]
  [[1. 1. 0.]
   [0. 1. 1.]
   [0. 0. 0.]]
  [[1. 0. 1.]
   [0. 0. 0.]
   [1. 1. 1.]]
  [[0. 0. 0.]
   [0. 0. 0.]
   [1. 0. 1.]]
  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]
  [[0. 0. 0.]
   [0. 0. 1.]
   [1. 0. 1.]]]]
se_conv2_out.grad:
[[[[ 20.]]
  [[ 37.]]
  [[ 90.]]
  [[ 10.]]
  [[ 74.]]
  [[  0.]]
  [[117.]]
  [[207.]]]]
se_gap_out.grad:
[[[[ -44.99]]
  [[  89.98]]
  [[ -25.  ]]
  [[ -25.  ]]
  [[-119.99]]
  [[   0.  ]]
  [[  45.  ]]
  [[ -10.  ]]]]
se_x_se.grad:
[[[[ -5.    -5.    -5.  ]
   [ -5.    -5.    -5.  ]
   [ -5.    -5.    -5.  ]]
  [[ 10.    10.    10.  ]
   [ 10.    10.    10.  ]
   [ 10.    10.    10.  ]]
  [[ -2.78  -2.78  -2.78]
   [ -2.78  -2.78  -2.78]
   [ -2.78  -2.78  -2.78]]
  [[ -2.78  -2.78  -2.78]
   [ -2.78  -2.78  -2.78]
   [ -2.78  -2.78  -2.78]]
  [[-13.33 -13.33 -13.33]
   [-13.33 -13.33 -13.33]
   [-13.33 -13.33 -13.33]]
  [[  0.     0.     0.  ]
   [  0.     0.     0.  ]
   [  0.     0.     0.  ]]
  [[  5.     5.     5.  ]
   [  5.     5.     5.  ]
   [  5.     5.     5.  ]]
  [[ -1.11  -1.11  -1.11]
   [ -1.11  -1.11  -1.11]
   [ -1.11  -1.11  -1.11]]]]
se_x_skip.grad:
[[[[0.5 0.5 0. ]
   [0.  0.  0.5]
   [0.  0.5 0. ]]
  [[1.  0.  0. ]
   [1.  1.  0. ]
   [1.  1.  1. ]]
  [[1.  0.  0. ]
   [0.  1.  1. ]
   [1.  1.  1. ]]
  [[1.  1.  0. ]
   [0.  1.  1. ]
   [0.  0.  0. ]]
  [[0.5 0.  0.5]
   [0.  0.  0. ]
   [0.5 0.5 0.5]]
  [[0.  0.  0. ]
   [0.  0.  0. ]
   [0.  0.  0. ]]
  [[1.  0.  0. ]
   [0.  1.  1. ]
   [1.  1.  1. ]]
  [[0.  0.  0. ]
   [0.  0.  0. ]
   [0.  0.  0. ]]]]
se_x.grad:
[[[[ -4.5   -4.5   -5.  ]
   [ -5.    -5.    -4.5 ]
   [ -5.    -4.5   -5.  ]]
  [[ 11.    10.    10.  ]
   [ 11.    11.    10.  ]
   [ 11.    11.    11.  ]]
  [[ -1.78  -2.78  -2.78]
   [ -2.78  -1.78  -1.78]
   [ -1.78  -1.78  -1.78]]
  [[ -1.78  -1.78  -2.78]
   [ -2.78  -1.78  -1.78]
   [ -2.78  -2.78  -2.78]]
  [[-12.83 -13.33 -12.83]
   [-13.33 -13.33 -13.33]
   [-12.83 -12.83 -12.83]]
  [[  0.     0.     0.  ]
   [  0.     0.     0.  ]
   [  0.     0.     0.  ]]
  [[  6.     5.     5.  ]
   [  5.     6.     6.  ]
   [  6.     6.     6.  ]]
  [[ -1.11  -1.11  -1.11]
   [ -1.11  -1.11  -1.11]
   [ -1.11  -1.11  -1.11]]]]
x_res.grad:
[[[[   0.      7.      0.  ]
   [  17.99  -45.96   21.01]
   [  27.     59.97   20.  ]]
  [[   0.     14.01    0.  ]
   [  35.97    0.     78.  ]
   [  18.03    0.      4.02]]
  [[   0.    -14.01    0.  ]
   [ -35.97   22.98  -69.01]
   [ -27.02  -29.98  -13.01]]
  [[   0.      0.      0.  ]
   [   0.    -22.98   -8.99]
   [   8.99   29.98    8.99]]
  [[   0.      7.      0.  ]
   [  17.99   22.98   47.99]
   [   0.02  -29.98   -6.98]]
  [[   0.      7.      0.  ]
   [  17.99    0.     39.  ]
   [   9.01    0.      2.01]]
  [[   0.    -21.01    0.  ]
   [ -53.96   22.98 -108.01]
   [ -36.03  -29.98  -15.02]]
  [[   0.     14.01    0.  ]
   [  35.97   22.98   86.99]
   [   9.03  -29.98   -4.97]]]]
x_skip.grad:
[[[[1. 1. 0.]
   [0. 0. 1.]
   [0. 1. 0.]]
  [[1. 0. 0.]
   [1. 1. 0.]
   [1. 1. 1.]]
  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]
  [[1. 1. 0.]
   [0. 1. 1.]
   [0. 0. 0.]]
  [[1. 0. 1.]
   [0. 0. 0.]
   [1. 1. 1.]]
  [[0. 0. 0.]
   [0. 0. 0.]
   [1. 0. 1.]]
  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]
  [[0. 0. 0.]
   [0. 0. 1.]
   [1. 0. 1.]]]]
x.grad:
[[[[   1.      8.      0.  ]
   [  17.99  -45.96   22.01]
   [  27.     60.97   20.  ]]
  [[   1.     14.01    0.  ]
   [  36.97    1.     78.  ]
   [  19.03    1.      5.02]]
  [[   1.    -14.01    0.  ]
   [ -35.97   23.98  -68.01]
   [ -26.02  -28.98  -12.01]]
  [[   1.      1.      0.  ]
   [   0.    -21.98   -7.99]
   [   8.99   29.98    8.99]]
  [[   1.      7.      1.  ]
   [  17.99   22.98   47.99]
   [   1.02  -28.98   -5.98]]
  [[   0.      7.      0.  ]
   [  17.99    0.     39.  ]
   [  10.01    0.      3.01]]
  [[   1.    -21.01    0.  ]
   [ -53.96   23.98 -107.01]
   [ -35.03  -28.98  -14.02]]
  [[   0.     14.01    0.  ]
   [  35.97   22.98   87.99]
   [  10.03  -29.98   -3.97]]]]
</code>
</details>

<hr style="border-width: 5px">

<a name="reproduktion-a14"></a>
### Die Implementierung ermöglicht es ebenfalls, die Ergebnisse von Übungsaufgabe 14 zu reproduzieren.

---

*ResNet-v1-Bottleneck-Block erstellen und Gewichte laden:*

Der Übersicht halber sind die Gewichte bereits im `tui-dl4cv`-Paket definiert und können über die Funktion `apply_block_weights` geladen werden.

In [9]:
# Block-Objekt erzeugen
block = BottleneckBlock(
    n_channels_in=8,
    n_channels=2,
    r_se=None    # kein SE verwenden
)

# Gewichte laden (nur fuer oben definierten Block)
apply_block_weights(block)

---

*Forward Pass:*

In [10]:
# Eingabe definieren
x = torch.tensor(
    [[[[1, 1, 0],
       [0, 0, 0],
       [0, 1, 0]],
      [[1, 0, 0],
       [1, 1, 0],
       [0, 1, 0]],
      [[1, 0, 0],
       [0, 1, 0],
       [0, 1, 0]],
      [[1, 1, 0],
       [0, 1, 2],
       [0, 0, 0]],
      [[3, 0, 2],
       [0, 0, 1],
       [0, 1, 0]],
      [[0, 0, 0],
       [0, 0, 0],
       [1, 0, 1]],
      [[1, 0, 0],
       [0, 3, 0],
       [0, 1, 0]],
      [[0, 0, 0],
       [0, 0, 2],
       [1, 0, 1]]]],
    dtype=torch.float32, requires_grad=True
)

# Forward Propagation
y = block(x)

# Ergebnis ueberpruefen
interactive_check('resnet_forward')

conv1_act:
[[[[ 1.  1.  2.]
   [ 2. -9.  5.]
   [ 3. -1.  3.]]

  [[ 2. -3.  2.]
   [ 0.  3.  1.]
   [ 1.  1.  1.]]]]
conv1_out:
[[[[1. 1. 2.]
   [2. 0. 5.]
   [3. 0. 3.]]

  [[2. 0. 2.]
   [0. 3. 1.]
   [1. 1. 1.]]]]
conv2_act:
[[[[ -1.  -5.  -9.]
   [ -1.   0.  -6.]
   [  9. -18.  10.]]

  [[ -5.  -1.  -1.]
   [ -1.  -4.  10.]
   [ 12. -14.   8.]]]]
conv2_out:
[[[[ 0.  0.  0.]
   [ 0.  0.  0.]
   [ 9.  0. 10.]]

  [[ 0.  0.  0.]
   [ 0.  0. 10.]
   [12.  0.  8.]]]]
conv3_act:
[[[[  0.   0.   0.]
   [  0.   0.  20.]
   [-30.   0. -44.]]

  [[  0.   0.   0.]
   [  0.   0. -10.]
   [ 15.   0.  22.]]

  [[  0.   0.   0.]
   [  0.   0.  30.]
   [ 36.   0.  24.]]

  [[  0.   0.   0.]
   [  0.   0.  10.]
   [-15.   0. -22.]]

  [[  0.   0.   0.]
   [  0.   0. -20.]
   [ 30.   0.  44.]]

  [[  0.   0.   0.]
   [  0.   0.   0.]
   [  0.   0.   0.]]

  [[  0.   0.   0.]
   [  0.   0.  20.]
   [ 51.   0.  46.]]

  [[  0.   0.   0.]
   [  0.   0.  50.]
   [ 87.   0.  70.]]]]
res:
[[[[  0.   0.  

Überprüfung der Ausgabe 'y':


<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
    <br>
    <i>Ausgaben bei korrekter Implementierung:</i>
    <br>
    <code style="padding: 0">
conv1_act:
[[[[ 1.  1.  2.]
   [ 2. -9.  5.]
   [ 3. -1.  3.]]
  [[ 2. -3.  2.]
   [ 0.  3.  1.]
   [ 1.  1.  1.]]]]
conv1_out:
[[[[1. 1. 2.]
   [2. 0. 5.]
   [3. 0. 3.]]
  [[2. 0. 2.]
   [0. 3. 1.]
   [1. 1. 1.]]]]
conv2_act:
[[[[ -1.  -5.  -9.]
   [ -1.   0.  -6.]
   [  9. -18.  10.]]
  [[ -5.  -1.  -1.]
   [ -1.  -4.  10.]
   [ 12. -14.   8.]]]]
conv2_out:
[[[[ 0.  0.  0.]
   [ 0.  0.  0.]
   [ 9.  0. 10.]]
  [[ 0.  0.  0.]
   [ 0.  0. 10.]
   [12.  0.  8.]]]]
conv3_act:
[[[[  0.   0.   0.]
   [  0.   0.  20.]
   [-30.   0. -44.]]
  [[  0.   0.   0.]
   [  0.   0. -10.]
   [ 15.   0.  22.]]
  [[  0.   0.   0.]
   [  0.   0.  30.]
   [ 36.   0.  24.]]
  [[  0.   0.   0.]
   [  0.   0.  10.]
   [-15.   0. -22.]]
  [[  0.   0.   0.]
   [  0.   0. -20.]
   [ 30.   0.  44.]]
  [[  0.   0.   0.]
   [  0.   0.   0.]
   [  0.   0.   0.]]
  [[  0.   0.   0.]
   [  0.   0.  20.]
   [ 51.   0.  46.]]
  [[  0.   0.   0.]
   [  0.   0.  50.]
   [ 87.   0.  70.]]]]
res:
[[[[  0.   0.   0.]
   [  0.   0.  20.]
   [-30.   0. -44.]]
  [[  0.   0.   0.]
   [  0.   0. -10.]
   [ 15.   0.  22.]]
  [[  0.   0.   0.]
   [  0.   0.  30.]
   [ 36.   0.  24.]
  [[  0.   0.   0.]
   [  0.   0.  10.]
   [-15.   0. -22.]]
  [[  0.   0.   0.]
   [  0.   0. -20.]
   [ 30.   0.  44.]]
  [[  0.   0.   0.]
   [  0.   0.   0.]
   [  0.   0.   0.]]
  [[  0.   0.   0.]
   [  0.   0.  20.]
   [ 51.   0.  46.]]
  [[  0.   0.   0.]
   [  0.   0.  50.]
   [ 87.   0.  70.]]]]
z:
[[[[  1.   1.   0.]
   [  0.   0.  20.]
   [-30.   1. -44.]]
  [[  1.   0.   0.]
   [  1.   1. -10.]
   [ 15.   1.  22.]]
  [[  1.   0.   0.]
   [  0.   1.  30.]
   [ 36.   1.  24.]]
  [[  1.   1.   0.]
   [  0.   1.  12.]
   [-15.   0. -22.]]
  [[  3.   0.   2.]
   [  0.   0. -19.]
   [ 30.   1.  44.]]
  [[  0.   0.   0.]
   [  0.   0.   0.]
   [  1.   0.   1.]]
  [[  1.   0.   0.]
   [  0.   3.  20.]
   [ 51.   1.  46.]]
  [[  0.   0.   0.]
   [  0.   0.  52.]
   [ 88.   0.  71.]]]]
y:
[[[[ 1.  1.  0.]
   [ 0.  0. 20.]
   [ 0.  1.  0.]]
  [[ 1.  0.  0.]
   [ 1.  1.  0.]
   [15.  1. 22.]]
  [[ 1.  0.  0.]
   [ 0.  1. 30.]
   [36.  1. 24.]]
  [[ 1.  1.  0.]
   [ 0.  1. 12.]
   [ 0.  0.  0.]]
  [[ 3.  0.  2.]
   [ 0.  0.  0.]
   [30.  1. 44.]]
  [[ 0.  0.  0.]
   [ 0.  0.  0.]
   [ 1.  0.  1.]]
  [[ 1.  0.  0.]
   [ 0.  3. 20.]
   [51.  1. 46.]]
  [[ 0.  0.  0.]
   [ 0.  0. 52.]
   [88.  0. 71.]]]]
</code>
</details>

---

*Backward Pass:*

In [11]:
# Gradienten definieren
y_grad = torch.ones(y.shape)

# Backpass durchfuehren
y.backward(y_grad)

# Ergebnis ueberpruefen
interactive_check('resnet_backward')

z.grad:
[[[[1. 1. 0.]
   [0. 0. 1.]
   [0. 1. 0.]]

  [[1. 0. 0.]
   [1. 1. 0.]
   [1. 1. 1.]]

  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]

  [[1. 1. 0.]
   [0. 1. 1.]
   [0. 0. 0.]]

  [[1. 0. 1.]
   [0. 0. 0.]
   [1. 1. 1.]]

  [[0. 0. 0.]
   [0. 0. 0.]
   [1. 0. 1.]]

  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]

  [[0. 0. 0.]
   [0. 0. 1.]
   [1. 0. 1.]]]]
res.grad:
[[[[1. 1. 0.]
   [0. 0. 1.]
   [0. 1. 0.]]

  [[1. 0. 0.]
   [1. 1. 0.]
   [1. 1. 1.]]

  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]

  [[1. 1. 0.]
   [0. 1. 1.]
   [0. 0. 0.]]

  [[1. 0. 1.]
   [0. 0. 0.]
   [1. 1. 1.]]

  [[0. 0. 0.]
   [0. 0. 0.]
   [1. 0. 1.]]

  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]

  [[0. 0. 0.]
   [0. 0. 1.]
   [1. 0. 1.]]]]
x_res.grad:
[[[[   0.   13.    0.]
   [  30.  -78.   39.]
   [  51.  104.   38.]]

  [[   0.   26.    0.]
   [  60.    0.  138.]
   [  42.    0.   16.]]

  [[   0.  -26.    0.]
   [ -60.   39. -123.]
   [ -57.  -52.  -31.]]

  [[   0.    0.    0.]
   [   0.  -39.  -15.]
   [ 

Überprüfung des Gradienten von 'x':


<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
    <br>
    <i>Augaben bei korrekter Implementierung:</i>
    <br>
    <code style="padding: 0">
z.grad:
[[[[1. 1. 0.]
   [0. 0. 1.]
   [0. 1. 0.]]
  [[1. 0. 0.]
   [1. 1. 0.]
   [1. 1. 1.]]
  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]
  [[1. 1. 0.]
   [0. 1. 1.]
   [0. 0. 0.]]
  [[1. 0. 1.]
   [0. 0. 0.]
   [1. 1. 1.]]
  [[0. 0. 0.]
   [0. 0. 0.]
   [1. 0. 1.]]
  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]
  [[0. 0. 0.]
   [0. 0. 1.]
   [1. 0. 1.]]]]
res.grad:
[[[[1. 1. 0.]
   [0. 0. 1.]
   [0. 1. 0.]]
  [[1. 0. 0.]
   [1. 1. 0.]
   [1. 1. 1.]]
  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]
  [[1. 1. 0.]
   [0. 1. 1.]
   [0. 0. 0.]]
  [[1. 0. 1.]
   [0. 0. 0.]
   [1. 1. 1.]]
  [[0. 0. 0.]
   [0. 0. 0.]
   [1. 0. 1.]]
  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]
  [[0. 0. 0.]
   [0. 0. 1.]
   [1. 0. 1.]]]]
x_res.grad:
[[[[   0.   13.    0.]
   [  30.  -78.   39.]
   [  51.  104.   38.]]
  [[   0.   26.    0.]
   [  60.    0.  138.]
   [  42.    0.   16.]]
  [[   0.  -26.    0.]
   [ -60.   39. -123.]
   [ -57.  -52.  -31.]]
  [[   0.    0.    0.]
   [   0.  -39.  -15.]
   [  15.   52.   15.]]
  [[   0.   13.    0.]
   [  30.   39.   84.]
   [   6.  -52.   -7.]]
  [[   0.   13.    0.]
   [  30.    0.   69.]
   [  21.    0.    8.]]
  [[   0.  -39.    0.]
   [ -90.   39. -192.]
   [ -78.  -52.  -39.]]
  [[   0.   26.    0.]
   [  60.   39.  153.]
   [  27.  -52.    1.]]]]
x_skip.grad:
[[[[1. 1. 0.]
   [0. 0. 1.]
   [0. 1. 0.]]
  [[1. 0. 0.]
   [1. 1. 0.]
   [1. 1. 1.]]
  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]
  [[1. 1. 0.]
   [0. 1. 1.]
   [0. 0. 0.]]
  [[1. 0. 1.]
   [0. 0. 0.]
   [1. 1. 1.]]
  [[0. 0. 0.]
   [0. 0. 0.]
   [1. 0. 1.]]
  [[1. 0. 0.]
   [0. 1. 1.]
   [1. 1. 1.]]
  [[0. 0. 0.]
   [0. 0. 1.]
   [1. 0. 1.]]]]
x.grad:
[[[[   1.   14.    0.]
   [  30.  -78.   40.]
   [  51.  105.   38.]]
  [[   1.   26.    0.]
   [  61.    1.  138.]
   [  43.    1.   17.]]
  [[   1.  -26.    0.]
   [ -60.   40. -122.]
   [ -56.  -51.  -30.]]
  [[   1.    1.    0.]
   [   0.  -38.  -14.]
   [  15.   52.   15.]]
  [[   1.   13.    1.]
   [  30.   39.   84.]
   [   7.  -51.   -6.]]
  [[   0.   13.    0.]
   [  30.    0.   69.]
   [  22.    0.    9.]]
  [[   1.  -39.    0.]
   [ -90.   40. -191.]
   [ -77.  -51.  -38.]]
  [[   0.   26.    0.]
   [  60.   39.  154.]
   [  28.  -52.    2.]]]]
</code>
</details>

$_{_\text{Created for Deep Learning for Computer Vision (DL4CV)}}$